Source/Core/Core/PowerPC/JitLLVM/Jit.cpp

```diff
@@ -0,0 +1,295 @@
+			Type::getVoidTy(con),
+			Type::getInt32Ty(con));
+
+		bound_func = Function::Create(type, Function::ExternalLinkage, opcodeInfo->opname, m_cur_mod->GetModule());
+		m_cur_binder->Bind(bound_func, (void*)instr);
+	}
+
+	builder->CreateCall(bound_func, builder->getInt32(inst.hex));
+}
+
+void JitLLVM::Break(LLVMFunction* func, UGeckoInstruction inst)
+{
+	// XXX: Actually break!
+}
+
+void JitLLVM::DoNothing(LLVMFunction* func, UGeckoInstruction inst)
+{
+}
+
+void JitLLVM::Jit(u32 em_address)
+{
+	if (em_address == 0)
+	{
+		Core::SetState(Core::CORE_PAUSE);
+		WARN_LOG(DYNA_REC, "ERROR: Compiling at 0. LR=%08x CTR=%08x", LR, CTR);
+	}
+
+	js.isLastInstruction = false;
+	js.firstFPInstructionFound = false;
+	js.blockStart = em_address;
+	js.fifoBytesThisBlock = 0;
+	js.downcountAmount = 0;
+	js.skipInstructions = 0;
+
+	int blockSize = code_buffer.GetSize();
+
+	if (SConfig::GetInstance().bEnableDebugging)
+	{
+		// Comment out the following to disable breakpoints (speed-up)
+		blockSize = 1;
+	}
+
+	u32 nextPC = em_address;
+	// Analyze the block, collect all instructions it is made of (including inlining,
+	// if that is enabled), reorder instructions for optimal performance, and join joinable instructions.
+	nextPC = analyzer.Analyze(em_address, &code_block, &code_buffer, blockSize);
+
+	legacy::PassManager PM;
+	PassManagerBuilder Builder;
+	Builder.OptLevel = 2;
+	Builder.populateModulePassManager(PM);
+	PM.add(createSLPVectorizerPass());
+	raw_ostream& out = outs();
+
+	if (m_debug_enabled)
+		PM.add(createPrintModulePass(out));
+
+	LLVMModule* mod = new LLVMModule(m_engine, em_address);
+	m_mods[em_address] = mod;
+
+	m_cur_mod = mod;
+	m_cur_func = mod->GetFunction();
+	m_cur_binder = mod->GetBinding();
+	m_mem_manager->SetNamedBinder(mod->GetNamedBinding());
+
+	// Build our GEP references first
+	m_cur_func->BuildGEPs();
+
+	PPCAnalyst::CodeOp *ops = code_buffer.codebuffer;
+
+	if (m_debug_enabled)
+	{
+		for (u32 i = 0; i < code_block.m_num_instructions; i++)
+		{
+			std::ostringstream ppc_disasm;
+			const PPCAnalyst::CodeOp &op = ops[i];
+			std::string opcode = GekkoDisassembler::Disassemble(op.inst.hex, op.address);
+			ppc_disasm << std::setfill('0') << std::setw(8) << std::hex << op.address;
+			ppc_disasm << " " << opcode << std::endl;
+			printf("%s", ppc_disasm.str().c_str());
+		}
+	}
+
+	IRBuilder<>* builder = m_cur_func->GetBuilder();
+
+	for (u32 i = 0; i < code_block.m_num_instructions; i++)
+	{
+		js.compilerPC = ops[i].address;
+		js.op = &ops[i];
+		js.instructionNumber = i;
+		const GekkoOPInfo *opinfo = ops[i].opinfo;
+		js.downcountAmount += opinfo->numCycles;
+
+		if (i == (code_block.m_num_instructions - 1))
+		{
+			// WARNING - cmp->branch merging will screw this up.
+			js.isLastInstruction = true;
+		}
+
+		if (!ops[i].skip)
+		{
+			if ((opinfo->flags & FL_USE_FPU) && !js.firstFPInstructionFound)
+			{
+				BasicBlock* disabled_block = m_cur_func->CreateBlock("fp_disabled");
```

Comment:
While we're thinking of reimplementing our block cache. What do you think about adding additional bits? eg we can recompile based on the information if the FPU is enabled.